# Numerische Differentiation

Die Ableitung einer Funktion \\(f:\mathbb{R} \rightarrow \mathbb{R}\\) an einer Stelle \\(x_0\\) ist definiert über den Grenzwert

\\[ f'(x_0) = \lim_{h\rightarrow0} \frac{f(x_0+h) - f(x_0)}{h}. \\]

Falls man weder über Ableitungsregeln noch über obigen Grenzwert die Ableitung analytisch berechnen kann, kann man die Ableitung numerisch approximieren, indem man den Differenzenquotient für einen festen (kleinen) Wert von \\(h\\) auswertet, anstatt den Grenzwert \\(h\rightarrow0\\) zu bilden.

Wir wollend diese Approximation anhand der Funktion \\(f(x) = \frac12 e^{x^2}\\) und der Stelle \\(x_0=1\\) testen. Diese Funktion kann man natürlich auch einfach von Hand ableiten als \\(f'(x) = xe^{x^2} \\), sodass wir für diese Funktion unsere Approximationen mit der exakten Ableitung vergleichen können.

Zunächst verwenden wir den Wert \\(h=10^{-6}\\) und vergleichen die Approximation mit der exakten Ableitung \\(f'(1) = e\\).

In [1]:
import numpy as np

In [13]:
x_0 = 1
h = 10**(-6)

approx = 1/h*( 0.5*np.exp((x_0+h)**2) - 0.5*np.exp(x_0**2) )
print('Approximation:', approx)
print('Approximationsfehler:', abs(approx - x_0 * np.exp(x_0**2)) )

Approximation: 2.7182859059227127
Approximationsfehler: 4.077463667595538e-06


Nun wollen wir beobachten, wie sich der Approximationsfehler für verschiedene Werte von \\(h\\) entwickelt. Dazu betrachten wir die Werte \\( h = 10^{-j}, j = 1,2,...,20 \\) und geben für jeden Wert von \\(h\\) den Approximationsfehler aus, verwenden also eine `for`-Schleife!

In [35]:
x_0 = 1

# oder: for j in [1,2,3,4,5,6,78,9,10,11,12,13,14,15,16,17,18,19,20]:
for j in range(1,21):
    h = 10**(-j)
    approx = 1/h*( 0.5*np.exp((x_0+h)**2) - 0.5*np.exp(x_0**2) )
    print('h = ',h,'--> Approximationsfehler:', abs( approx - x_0 * np.exp(x_0**2)) )

h =  0.1 --> Approximationsfehler: 0.45773229199085197
h =  0.01 --> Approximationsfehler: 0.04123161394521491
h =  0.001 --> Approximationsfehler: 0.004081957519520341
h =  0.0001 --> Approximationsfehler: 0.000407787582335839
h =  1e-05 --> Approximationsfehler: 4.0774708909729895e-05
h =  1e-06 --> Approximationsfehler: 4.077463667595538e-06
h =  1e-07 --> Approximationsfehler: 4.108411064684958e-07
h =  1e-08 --> Approximationsfehler: 1.560170970194008e-08
h =  1e-09 --> Approximationsfehler: 2.1544185369037905e-07
h =  1e-10 --> Approximationsfehler: 1.5477094836846561e-06
h =  1e-11 --> Approximationsfehler: 1.0429493680685908e-05
h =  1e-12 --> Approximationsfehler: 0.00021026963811321409
h =  1e-13 --> Approximationsfehler: 0.00045586417666187984
h =  1e-14 --> Approximationsfehler: 0.012866812118839999
h =  1e-15 --> Approximationsfehler: 0.3903426404913928
h =  1e-16 --> Approximationsfehler: 2.718281828459045
h =  1e-17 --> Approximationsfehler: 2.718281828459045
h =  1e-18 

Beobachtung: Der Approximationsfehler wird zunächst immer kleiner, geht bis auf die Größenordnung \\(10^{-8}\\) runter, steigt dann aber wieder und stagniert letztendlich bei \\( 2.718281828459045 \approx e \\).

Warum?